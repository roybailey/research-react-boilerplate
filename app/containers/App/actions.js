/*
 * App Actions
 *
 * Actions change things in your application
 * Since this boilerplate uses a uni-directional data flow, specifically redux,
 * we have these actions which are the only way your application interacts with
 * your application state. This guarantees that your state is up to date and nobody
 * messes it up weirdly somewhere.
 *
 * To add a new Action:
 * 1) Import your constant
 * 2) Add a function like this:
 *    export function yourAction(var) {
 *        return { type: YOUR_ACTION_CONSTANT, var: var }
 *    }
 */

import {
  LOAD_REPOS,
  LOAD_REPOS_SUCCESS,
  LOAD_REPOS_ERROR,
  LOAD_TODOS,
  LOAD_TODOS_SUCCESS,
  LOAD_TODOS_ERROR,
} from './constants';

/**
 * Load the repositories, this action starts the request saga
 *
 * @return {object} An action object with a type of LOAD_REPOS
 */
export function loadRepos() {
  return {
    type: LOAD_REPOS,
  };
}

/**
 * Dispatched when the repositories are loaded by the request saga
 *
 * @param  {array} repos The repository data
 * @param  {string} username The current username
 *
 * @return {object}      An action object with a type of LOAD_REPOS_SUCCESS passing the repos
 */
export function reposLoaded(repos, username) {
  return {
    type: LOAD_REPOS_SUCCESS,
    repos,
    username,
  };
}

/**
 * Dispatched when loading the repositories fails
 *
 * @param  {object} error The error
 *
 * @return {object}       An action object with a type of LOAD_REPOS_ERROR passing the error
 */
export function repoLoadingError(error) {
  return {
    type: LOAD_REPOS_ERROR,
    error,
  };
}

// ------------------------------------------------------------
// TODOS
// ------------------------------------------------------------

/**
 * Load the todos, this action starts the request saga
 *
 * @return {object} An action object with a type of LOAD_TODOS
 */
export function loadTodos() {
  return {
    type: LOAD_TODOS,
  };
}

/**
 * Dispatched when the todos are loaded by the request saga
 *
 * @param  {array} todos The todos data
 * @param  {string} category The current category
 *
 * @return {object}      An action object with a type of LOAD_TODOS_SUCCESS passing the todos
 */
export function todosLoaded(todos, category) {
  return {
    type: LOAD_TODOS_SUCCESS,
    todos,
    category,
  };
}

/**
 * Dispatched when loading the todos fails
 *
 * @param  {object} error The error
 *
 * @return {object}       An action object with a type of LOAD_TODOS_ERROR passing the error
 */
export function todosLoadingError(error) {
  return {
    type: LOAD_TODOS_ERROR,
    error,
  };
}
