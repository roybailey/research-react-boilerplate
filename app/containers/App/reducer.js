/*
 * AppReducer
 *
 * The reducer takes care of our data. Using actions, we can change our
 * application state.
 * To add a new action, add it to the switch statement in the reducer function
 *
 * Example:
 * case YOUR_ACTION_CONSTANT:
 *   return state.set('yourStateVariable', true);
 */

import { fromJS } from 'immutable';

import {
  LOAD_REPOS_SUCCESS,
  LOAD_REPOS,
  LOAD_REPOS_ERROR,
  LOAD_TODOS_SUCCESS,
  LOAD_TODOS,
  LOAD_TODOS_ERROR,
} from './constants';

// The initial state of the App
const initialState = fromJS({
  loading: false,
  error: false,
  currentUser: false,
  userData: {
    repositories: false,
  },
  todoData: {
    todos: false,
  },
});

function appReducer(state = initialState, action) {
  switch (action.type) {
    case LOAD_REPOS:
      return state
        .set('loading', true)
        .set('error', false)
        .setIn(['userData', 'repositories'], false);
    case LOAD_REPOS_SUCCESS:
      return state
        .setIn(['userData', 'repositories'], action.repos)
        .set('loading', false)
        .set('currentUser', action.username);
    case LOAD_REPOS_ERROR:
      return state.set('error', action.error).set('loading', false);

    case LOAD_TODOS:
      return state
        .setIn(['todoData', 'loading'], true)
        .setIn(['todoData', 'error'], false)
        .setIn(['todoData', 'todos'], false);
    case LOAD_TODOS_SUCCESS:
      return state
        .setIn(['todoData', 'loading'], false)
        .setIn(['todoData', 'category'], action.category)
        .setIn(['todoData', 'todos'], action.todos);
    case LOAD_TODOS_ERROR:
      return state
        .setIn(['todoData', 'loading'], false)
        .setIn(['todoData', 'error'], action.error);
    default:
      return state;
  }
}

export default appReducer;
