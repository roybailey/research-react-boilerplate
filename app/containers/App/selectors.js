/**
 * The global state selectors
 */

import { createSelector } from 'reselect';

const selectGlobal = state => state.get('global');

const selectRoute = state => state.get('route');

const makeSelectCurrentUser = () =>
  createSelector(selectGlobal, globalState => globalState.get('currentUser'));

const makeSelectLoading = () =>
  createSelector(selectGlobal, globalState => globalState.get('loading'));

const makeSelectError = () =>
  createSelector(selectGlobal, globalState => globalState.get('error'));

const makeSelectRepos = () =>
  createSelector(selectGlobal, globalState =>
    globalState.getIn(['userData', 'repositories']),
  );

const makeSelectTodosByCurrentStatus = () =>
  createSelector(selectGlobal, globalState => {
    const status = globalState.getIn(['todoData', 'category']);
    const hasData = globalState.getIn(['todoData', 'todos']);
    return hasData
      ? hasData.filter(todo => todo.status === status.value)
      : hasData;
  });

const makeSelectTodos = () =>
  createSelector(selectGlobal, globalState =>
    globalState.getIn(['todoData', 'todos']),
  );

const makeSelectLocation = () =>
  createSelector(selectRoute, routeState => routeState.get('location').toJS());

export {
  selectGlobal,
  makeSelectCurrentUser,
  makeSelectLoading,
  makeSelectError,
  makeSelectRepos,
  makeSelectTodosByCurrentStatus,
  makeSelectTodos,
  makeSelectLocation,
};
