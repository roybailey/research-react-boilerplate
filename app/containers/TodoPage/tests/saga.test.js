/* eslint-disable redux-saga/yield-effects */
/**
 * Tests for HomePage sagas
 */

import { put, takeLatest } from 'redux-saga/effects';

import { LOAD_TODOS } from 'containers/App/constants';
import { todosLoaded, todosLoadingError } from 'containers/App/actions';

import todoData, { getTodos } from '../saga';

const category = { value: 'work' };

describe('getTodos Saga', () => {
  let getTodosGenerator;

  // We have to test twice, once for a successful load and once for an unsuccessful one
  // so we do all the stuff that happens beforehand automatically in the beforeEach
  beforeEach(() => {
    getTodosGenerator = getTodos();

    const selectDescriptor = getTodosGenerator.next().value;
    expect(selectDescriptor).toMatchSnapshot();

    const callDescriptor = getTodosGenerator.next(category).value;
    expect(callDescriptor).toMatchSnapshot();
  });

  it('should dispatch the todosLoaded action if it requests the data successfully', () => {
    const response = [
      {
        name: 'First todo',
      },
      {
        name: 'Second todo',
      },
    ];
    const putDescriptor = getTodosGenerator.next(response).value;
    expect(putDescriptor).toEqual(put(todosLoaded(response, category)));
  });

  it('should call the todoLoadingError action if the response errors', () => {
    const response = new Error('Some error');
    const putDescriptor = getTodosGenerator.throw(response).value;
    expect(putDescriptor).toEqual(put(todosLoadingError(response)));
  });
});

describe('todosDataSaga Saga', () => {
  const todosDataSaga = todoData();

  it('should start task to watch for LOAD_TODOS action', () => {
    const takeLatestDescriptor = todosDataSaga.next().value;
    expect(takeLatestDescriptor).toEqual(takeLatest(LOAD_TODOS, getTodos));
  });
});
