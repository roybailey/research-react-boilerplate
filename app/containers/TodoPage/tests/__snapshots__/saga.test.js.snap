// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getTodos Saga should call the todoLoadingError action if the response errors 1`] = `
Object {
  "@@redux-saga/IO": true,
  "SELECT": Object {
    "args": Array [],
    "selector": [Function],
  },
}
`;

exports[`getTodos Saga should call the todoLoadingError action if the response errors 2`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "/api/todo?category=work",
    ],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`getTodos Saga should dispatch the todosLoaded action if it requests the data successfully 1`] = `
Object {
  "@@redux-saga/IO": true,
  "SELECT": Object {
    "args": Array [],
    "selector": [Function],
  },
}
`;

exports[`getTodos Saga should dispatch the todosLoaded action if it requests the data successfully 2`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "/api/todo?category=work",
    ],
    "context": null,
    "fn": [Function],
  },
}
`;
